### PR TITLE
Added delete and massDelete methods to menu listing

### PR DESCRIPTION
**Pull request by kiber**
DEV-69083

### CODE REVIEW

File: Controller/Adminhtml/Menu/MassDelete.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+     * @return Redirect
+     * @throws LocalizedException|Exception
+     */
+    public function execute()
+    {
+        $collection = $this->filter->getCollection($this->collectionFactory->create());
+        $collectionSize = $collection->getSize();
+
+        foreach ($collection as $menu) {
+            $menu->delete();
```
Comment (jochhop) [inline Controller/Adminhtml/Menu/MassDelete.php:68]: Please use `\Snowdog\Menu\Api\MenuRepositoryInterface`'s `delete` as `delete` from `AbstractModel` is deprecated.

File: Controller/Adminhtml/Menu/MassDelete.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    /**
+     * Execute action
+     *
+     * @return Redirect
+     * @throws LocalizedException|Exception
+     */
+    public function execute()
+    {
+        $collection = $this->filter->getCollection($this->collectionFactory->create());
```
Comment (jochhop) [inline Controller/Adminhtml/Menu/MassDelete.php:64]: I'd move this into a private method, so we can just call it eg. `$this->getMenuCollection()`

File: Controller/Adminhtml/Menu/MassDelete.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        $collectionSize = $collection->getSize();
```
Comment (jochhop) [inline Controller/Adminhtml/Menu/MassDelete.php:65]: I'd move this after the `foreach`, this way if something goes wrong during menu removal this variable does not initialize

File: Controller/Adminhtml/Menu/MassDelete.php (hunk 1 above)
Comment (kiber) [reply Controller/Adminhtml/Menu/MassDelete.php:68]: Done

File: Controller/Adminhtml/Menu/MassDelete.php (hunk 2 above)
Comment (kiber) [reply Controller/Adminhtml/Menu/MassDelete.php:64]: Done

File: Controller/Adminhtml/Menu/MassDelete.php (hunk 3 above)
Comment (kiber) [reply Controller/Adminhtml/Menu/MassDelete.php:65]: I have tested your suggestion but it returns 0 after `foreach` loop.

Comment (jochhop) [reply Controller/Adminhtml/Menu/MassDelete.php:75]: Yeah... my bad, it's removing it from db and from collection, sorry 🤦

File: Controller/Adminhtml/Menu/Delete.php [hunk 4]
```diff
@@ -53,19 +81,22 @@ public function __construct(
...
     public function execute()
     {
-        $id = $this->getRequest()->getParam('id');
+        $menuId = (int) $this->getRequest()->getParam('id');
 
         try {
-            $menu = $this->menuRepository->getById($id);
-            $this->menuRepository->deleteById($id);
+            $menu = $this->menuRepository->getById($menuId);
+            $this->menuRepository->deleteById($menuId);
```
Comment (jochhop) [inline Controller/Adminhtml/Menu/Delete.php:93]: I think we could use `$this->menuRepository->delete($menu)`, this way we avoid loading a menu twice

File: Ui/Component/Listing/Column/MenuList/PageActions.php [hunk 5]
```diff
@@ -1,26 +1,93 @@
...
-                $name = $this->getData("name");
-                $id = "X";
-                if (isset($item["menu_id"])) {
-                    $id = $item["menu_id"];
+        if (isset($dataSource['data']['items'])) {
+            foreach ($dataSource['data']['items'] as & $item) {
+                if (isset($item['menu_id'])) {
+                    $name = $this->getData('name');
+                    $item[$name] = [
+                        'edit' => [
```
Comment (jochhop) [inline Ui/Component/Listing/Column/MenuList/PageActions.php:64]: This method seems pretty long, what do you think if we extract `edit` and `delete` buttons generation into 2 different private methods so this part will look like:

```php
// ...
$item[$name] = [
    'edit' => $this->getEditButton($item['menu_id']),
    'delete' => $this->getDeleteButton($item['menu_id']),    
]
// ...
```

File: Controller/Adminhtml/Menu/MassDelete.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+     * Authorization level of a basic admin session
+     *
+     * @see _isAllowed()
+     */
+    const ADMIN_RESOURCE = 'Snowdog_Menu::menus';
+
+    /**
+     * @var Filter
+     */
+    protected $filter;
```
Comment (jochhop) [inline Controller/Adminhtml/Menu/MassDelete.php:36]: Is there any reason why these properties are set as `protected`, it looks like they are not going to be overriden/used in a child class, I think we could switch them to `private` instead.

Comment (kiber) [reply Controller/Adminhtml/Menu/MassDelete.php:36]: There are other properties that are `protected`, that is why I left it also as `protected`. Now I changed it to `private`

File: Ui/Component/Listing/Column/MenuList/PageActions.php (hunk 5 above)
Comment (kiber) [reply Ui/Component/Listing/Column/MenuList/PageActions.php:64]: Done

File: Controller/Adminhtml/Menu/Delete.php (hunk 4 above)
Comment (kiber) [reply Controller/Adminhtml/Menu/Delete.php:93]: Done

**Review by jochhop — DISMISSED** [review]

File: Controller/Adminhtml/Menu/Delete.php [hunk 7]
```diff
@@ -53,19 +81,22 @@ public function __construct(
...
      */
     public function execute()
     {
-        $id = $this->getRequest()->getParam('id');
+        $menuId = (int) $this->getRequest()->getParam('id');
 
         try {
-            $menu = $this->menuRepository->getById($id);
-            $this->menuRepository->deleteById($id);
+            $menu = $this->menuRepository->getById($menuId);
```
Comment (TM18) [inline Controller/Adminhtml/Menu/Delete.php:92]: `->getById($menuId)` throws a `NoSuchEntityException`, I wonder if we should catch it as well, same as with `CouldNotDeleteException`. I see both extend the `LocalizedException`, maybe catch this one only, what do you think?

File: Controller/Adminhtml/Menu/MassDelete.php [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+     * @return Redirect
+     * @throws LocalizedException|Exception
+     */
+    public function execute()
+    {
+        $collection = $this->getMenuCollection();
+        $collectionSize = $collection->getSize();
+
+        foreach ($collection as $menu) {
+            $this->menuRepository->delete($menu);
```
Comment (TM18) [inline Controller/Adminhtml/Menu/MassDelete.php:78]: Shouldn't we catch `CouldNotDeleteException` here as well?

Comment (kiber) [reply Controller/Adminhtml/Menu/MassDelete.php]: Done

File: Controller/Adminhtml/Menu/Delete.php (hunk 7 above)
Comment (kiber) [reply Controller/Adminhtml/Menu/Delete.php:93]: Added another line to catch `NoSuchEntityException` exception. I think this way more clear than using one `LocalizedException` exception.

**Review by jochhop — APPROVED** [review]

**Pull request merged**